sdk/src/trace/batch_span_processor.cc

```diff
@@ -0,0 +1,224 @@
+{
+  return exporter_->MakeRecordable();
+}
+
+void BatchSpanProcessor::OnStart(Recordable &span) noexcept
+{
+  // no-op
+  (void)span;
+}
+
+void BatchSpanProcessor::OnEnd(std::unique_ptr<Recordable> &&span) noexcept
+{
+  if (is_shutdown_.load())
+  {
+    return;
+  }
+
+  if (buffer_->Add(span) == false)
+  {
+    return;
+  }
+
+  // If the queue gets at least half full a preemptive notification is
+  // sent to the worker thread to start a new export cycle.
+  if (buffer_->size() >= max_queue_size_ / 2)
+  {
+    // signal the worker thread
+    cv_.notify_one();
+  }
+}
+
+void BatchSpanProcessor::ForceFlush(std::chrono::microseconds timeout) noexcept
+{
+  if (is_shutdown_.load())
+  {
+    return;
+  }
+
+  is_force_flush_ = true;
+
+  // Keep attempting to wake up the worker thread
+  while (is_force_flush_.load() == true)
+  {
+    cv_.notify_one();
+  }
+
+  // Now wait for the worker thread to signal back from the Export method
+  std::unique_lock<std::mutex> lk(force_flush_cv_m_);
+  while (is_force_flush_notified_.load() == false)
+  {
+    force_flush_cv_.wait(lk);
+  }
+
+  // Notify the worker thread
+  is_force_flush_notified_ = false;
+}
+
+void BatchSpanProcessor::DoBackgroundWork()
+{
+  auto timeout = schedule_delay_millis_;
+
+  while (true)
+  {
+    // If we already have max_export_batch_size_ spans in the buffer, better to export them
+    // now
+    if (buffer_->size() < max_export_batch_size_)
+    {
+      // In case of spurious wake up, we export only if we have atleast one span
+      // in the batch. This is acceptable because batching is a best mechanism
+      // effort here.
+      do
+      {
+        std::unique_lock<std::mutex> lk(cv_m_);
+
+        cv_.wait_for(lk, std::chrono::milliseconds(timeout));
+
+        // If shutdown has been invoked, break out of the loop
+        // and drain out the queue.
+        if (is_shutdown_.load() == true)
+        {
+          is_shutdown_ = false;
+          // ensure the main thread has been notified
+          while (is_shutdown_.load() == false)
+            ;
+          DrainQueue();
+          return;
+        }
+
+        if (is_force_flush_.load() == true)
+        {
+          break;
+        }
+      } while (buffer_->empty() == true);
+    }
+
+    // Get the value of is_force_flush_ to check whether this export is
+    // a result of a ForceFlush call. This flag is propagated to the Export
+    // method as well.
+    bool was_force_flush_called = is_force_flush_.load();
+
+    // Set the flag back to false to notify main thread in case ForceFlush
+    // was invoked.
+    if (is_force_flush_.load() == true)
+    {
+      is_force_flush_ = false;
+    }
+
+    auto start = std::chrono::steady_clock::now();
+    Export(was_force_flush_called);
+    auto end      = std::chrono::steady_clock::now();
+    auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(end - start);
+
+    timeout = schedule_delay_millis_ - duration;
```

Comment:
I think we can get rid of this calculation and just go with waiting for the whole `schedule_delay_millis_` in each iteration. This would be more in line with the spec, which says:

> the delay interval in milliseconds _between_ two consecutive exports

Reply:
I was under the impression that it meant the delay interval between when two exports essentially *began*.
I believe [python](https://github.com/open-telemetry/opentelemetry-python/blob/master/opentelemetry-sdk/src/opentelemetry/sdk/trace/export/__init__.py#L186) has the same logic too.

Reply:
That's a good point. I had a look at [Java](https://github.com/open-telemetry/opentelemetry-java/blob/master/sdk/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java#L215), which doesn't have it.

I was drawn to the simpler solution, but I leave this up to you to decide.